spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/predicate/ReadBodyPredicateFactory.java

```diff
@@ -70,39 +77,48 @@ public ReadBodyPredicateFactory(ServerCodecConfigurer codecConfigurer) {
 				try {
 					boolean test = config.predicate.test(cachedBody);
 					exchange.getAttributes().put(TEST_ATTRIBUTE, test);
+					return Mono.just(test);
 				} catch(ClassCastException e) {
 					if(LOGGER.isDebugEnabled()) {
 						LOGGER.debug("Predicate test failed because class in predicate does not match the cached body object",
 								e);
 					}
 				}
-				modifiedBody = Mono.just(cachedBody);
+				return Mono.just(false);
 			} else {
-				ServerRequest serverRequest = new DefaultServerRequest(exchange);
-				// TODO: flux or mono
-				modifiedBody = serverRequest.bodyToMono(inClass)
-						// .log("modify_request_mono", Level.INFO)
-						.flatMap(body -> {
-							// TODO: migrate to async
-							exchange.getAttributes().put(CACHE_REQUEST_BODY_OBJECT_KEY, body);
-							boolean test = config.predicate.test(body);
-							exchange.getAttributes().put(TEST_ATTRIBUTE, test);
-							return Mono.just(body);
-						});
-			}
-			BodyInserter bodyInserter = BodyInserters.fromPublisher(modifiedBody, inClass);
-			CachedBodyOutputMessage outputMessage = new CachedBodyOutputMessage(exchange,
-					exchange.getRequest().getHeaders());
-			return bodyInserter.insert(outputMessage, new BodyInserterContext())
-					// .log("modify_request", Level.INFO)
-					.then(Mono.defer(() -> {
-						boolean test = (Boolean) exchange.getAttributes()
-								.getOrDefault(TEST_ATTRIBUTE, Boolean.FALSE);
-						exchange.getAttributes().remove(TEST_ATTRIBUTE);
-						exchange.getAttributes().put(CACHED_REQUEST_BODY_KEY,
+				Flux<DataBuffer> origBody = exchange.getRequest().getBody().flatMap(dataBuffer -> {
```

Comment:
Does this work for large objects? It's a `Flux<DataBuffer>`, the object might be spread across more than one. From what I can tell, it looks like it's only converting one `DataBuffer` at a time. 

I think rather than flatMap the body, couldn't you move this code out of flatmap and work on the `Flux<DataBuffer>` at `ReactiveHttpInputMessage`?

Might need some advice from @rstoyanchev

Reply:
Indeed, passing chunks individually into a reader makes no sense. I don't get the point of replacing `ServerRequest` actually.

Reply:
@spencergibb @rstoyanchev that makes sense.  

What I want is to read the body once but end up with 2 copies of the `Flux<DataBuffer>`, one for to create a new request from (since we already read the body and other parts of the Gateway will need to read it again) and a second to convert the body to an object and perform the predicate test on.  I am not sure the best way to do that.  Maybe `Flux.groupBy` or` Flux.repeat` ?